HIP/hip-1195.md

```diff
@@ -0,0 +1,834 @@
+
+  /**
+   * If positive, the index of the first hook installed on this account.
+   */
+  uint64 first_hook_index = 37;
+}
+```
+
+For a successful such `CryptoCreate` or `CryptoUpdate`, the indexes of the newly installed hooks will appear in the
+legacy record `TransactionReceipt` if record streams are still enabled.
+
+Now we need to let a `CryptoTransfer` reference such a hook. For this we extend the `AccountAmount` and `NftTransfer`
+messages used in the `CryptoTransferTransactionBody`.
+
+```protobuf
+message AccountAmount {
+  // ...
+  /**
+   * If set, a call to a hook of type `ACCOUNT_ALLOWANCE_HOOK` installed on
+   * accountID that must succeed for the transaction to occur.
+   */
+  HookCall allowance_hook_call = 4;
+}
+
+message NftTransfer {
+  // ...
+  /**
+   * If set, a call to a hook of type `ACCOUNT_ALLOWANCE_HOOK` installed on
+   * senderAccountID that must succeed for the transaction to occur.
+   */
+  HookCall sender_allowance_hook_call = 5;
+
+  /**
+   * If set, a call to a hook of type `ACCOUNT_ALLOWANCE_HOOK` installed on
+   * receiverAccountID that must succeed for the transaction to occur.
+   */
+  HookCall receiver_allowance_hook_call = 6;
+}
+```
+
+Note that `NftTransfer` supports both sender and receiver transfer allowance hooks, since the transaction may
+need to use the receiver hook to satisfy a `receiver_sig_required=true` setting.
+
+### The transfer allowance ABI
+
+The account allowance EVM hook ABI is as follows,
+
+```solidity
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.4.9 <0.9.0;
+pragma experimental ABIEncoderV2;
+
+import './IHederaTokenService.sol';
+
+/// The interface for an account allowance EVM hook.
+interface IHieroAccountAllowanceEvmHook {
+    /// Combines HBAR and HTS asset transfers.
+    struct Transfers {
+        /// The HBAR transfers
+        IHederaTokenService.TransferList hbar;
+        /// The HTS token transfers
+        IHederaTokenService.TokenTransferList[] tokens;
+    }
+
+    /// Combines the full proposed transfers for a Hiero transaction,
+    /// including both its direct transfers and the implied HIP-18
+    /// custom fee transfers.
+    struct ProposedTransfers {
+        /// The transaction's direct transfers
+        Transfers direct;
+        /// The transaction's assessed custom fees
+        Transfers customFee;
+    }
+
+    /// Decides if the proposed transfers are allowed, optionally in
+    /// the presence of additional context encoded by the transaction
+    /// payer in the extra args.
+    /// @param installer The address of the installing account for which the hook is being executed
+    /// @param proposedTransfers The proposed transfers
+    /// @param args The extra arguments
+    /// @return true If the proposed transfers are allowed, false or revert otherwise
+    function allow(
+       address installer,
+       ProposedTransfers memory proposedTransfers,
+       bytes memory args
+    ) external payable returns (bool);
+}
```

Comment:
I'm not sure the ABI is coherent with the current proto definition.

If we don't modify the proto messages by moving the `HookCall` outside of the single transfer operation, it seems to me that the hook will not receive data for *all transfers*, but only for the transfer in which the hook is involved.

My understanding is that, in the case of multiple transfers, the hook will be executed multiple times.
Is that correct? If so, the ABI interface and the examples below should be updated to reflect this.

Otherwise, we should clarify what happens when I define different hooks for different transfers in the same transaction. Would I receive them all? Will the hook system contract group all transfers related to the same hook?

Reply:
Thanks. Just to be sure: if any of the hook authorization fails, the entire transaction fails, right?

Reply:
Yes, exactly!